### PR TITLE
[WIP] Various bugs in passes

### PR DESCRIPTION
**Pull request by mbrookhart**
This adds regression tests for a number of bugs uncovered while testing ONNX Object  Detection models, and adds regression tests for the fixes

@tqchen

### CODE REVIEW

**Review by tqchen — APPROVED** [review]

File: src/relay/op/tensor/transform.cc
Comment (masahi) [inline src/relay/op/tensor/transform.cc:1079]: I hope we could come up with a way to avoid this. This would hurt performance on hummingbird workload and possibly other models.

Comment (masahi) [reply src/relay/op/tensor/transform.cc:1079]: For example, maybe we can use `annotation.stop_fusion` when we encounter take + dynamic, solving this problem at the frontend.

Comment (mbrookhart) [reply src/relay/op/tensor/transform.cc:1079]: Good thing I have a typo in CI.

I'm not sure I see a clean way to do this in the frontends, it demands we already have infer_type run to check for dynamic inputs.

Maybe we write a pass that selectively stops fusion on certain ops under certain conditions?

File: src/relay/transforms/fold_scale_axis.cc
Comment (masahi) [inline src/relay/transforms/fold_scale_axis.cc:246]: cc @kevinthesun This might be of interest to you

File: src/relay/op/tensor/transform.cc
Comment (masahi) [reply src/relay/op/tensor/transform.cc:1079]: Yes a new pass that inserts `stop_fusion` sounds good. I can work on this. We can make `take` opaque for now until I have that new pass ready.

Anyway I think graph runtime shouldn't be affected by the issue of `take` + dynamic, so `take` should be injective eventually.

Comment (masahi) [reply src/relay/op/tensor/transform.cc:1079]: @mbrookhart Can you add `TODO(masahi)` there?

Comment (mbrookhart) [reply src/relay/op/tensor/transform.cc:1079]: @masahi want to request changes so this doesn't merge while we chat about it?

**Review by masahi — CHANGES_REQUESTED** [review]
Block merging until we settle on the `take` fusion issue

File: src/relay/op/tensor/transform.cc
Comment (masahi) [reply src/relay/op/tensor/transform.cc:1079]: sure done. But I think it is good to go after adding a comment on why we make this change temporary

Comment (mbrookhart) [reply src/relay/op/tensor/transform.cc:1079]: I think the other option is to edit the relay take function that creates the op. We could remove the normalization that causes this problem from the take kernel in topi, and do in it relay with select/shape_of, but that might end up causing some performance degradation, it's hard to predict.

**Comment by tqchen** [comment]
@mbrookhart perhaps we can separate out the other things (fold scale axis and bound) into separate PRs so we can merge them in first. It would also makes the PR itself more informative.

**Comment by mbrookhart** [comment]
@tqchen @masahi That sounds good, I'll split the other parts off into a new PR, and we can wait on `take` until we have a more robust solution.

**Comment by masahi** [comment]
@mbrookhart maybe you can summarize what the issue is with `take` + dynamic (for me + others to comment)?  I haven't looked into `take` topi compute in detail, so I haven't understood what's wrong with it or what do you mean by "normalization".

I also want to know why this could lead to infinite loop inside `rewrite_simplify` (visiting the same node over and over again) @tqchen

**Comment by mbrookhart** [comment]
I've been working on other things, sorry for the delay on this.

take with dynamic input shapes is a bit of an issue because take folds the input shape into the compute:
https://github.com/apache/tvm/blob/c58b20bbd8e6ac7128e82994624d922ee6b8a069/include/tvm/topi/transform.h#L760

If the input tensor to take is fused into other ops, that input shape maybe removed in the fusion, and then we'll have a reference to an input dimension that no longer exists, so the fused function will fail to compile.

As @masahi and I discussed above, there are two main ways to get around this: We can write a pass that breaks the fusion at an earlier step, or we can remove the `clip` and `wrap`modes from the topi op and raise that logic into relay. I have a branch where I've done the second and it passes forward tests, but I'm having some issues getting gradient tests and ADT tests working.

A third option I've thought of in the meantime is to do the raising of the logic in a legalization pass only for dynamic input shapes. This would allow us to remove the dynamic issue but keep the current code everywhere else.

Any thoughts on the best way to proceed?

**Pull request closed without merging**
